mysql-test/suite/rpl/t/rpl_semi_sync_master_disable_with_slave.test

```diff
@@ -0,0 +1,41 @@
+# MDEV-36359:
+# Binlog Dump thread crashes when reverting to async after Semi-Sync disabled.
+#
+# Assert behavior of turning Semi-Sync off on
+# the master when still connected to a slave
+
+--source include/have_binlog_format_mixed.inc # format-agnostic
+--source include/master-slave.inc
+
+--let $status_var= Rpl_semi_sync_master_status
+SELECT ID INTO @binlog_dump_tid
+  FROM information_schema.PROCESSLIST WHERE COMMAND = 'Binlog Dump';
+
+--echo # Control State
+SELECT STATE FROM information_schema.PROCESSLIST WHERE ID = @binlog_dump_tid;
+SHOW STATUS LIKE 'Rpl_semi_sync_master_clients';
+
+--echo # Disable Semi-Sync while the dump thread is connected
+SET @@GLOBAL.rpl_semi_sync_master_enabled = 0;
+--let $status_var_value= OFF
+--source include/wait_for_status_var.inc
+
+SELECT STATE FROM information_schema.PROCESSLIST WHERE ID = @binlog_dump_tid;
+SHOW STATUS LIKE 'Rpl_semi_sync_master_clients';
+
+--echo # Disconnect the slave and wait until the master's dump thread is done
+--connection slave
+STOP SLAVE;
+--connection master
+
+# MDEV-36359: 2013 'Lost connection to server during query'
+--let $wait_condition= SELECT COUNT(*)=0 FROM information_schema.PROCESSLIST WHERE ID = @binlog_dump_tid
+--source include/wait_condition.inc
```

Comment:
Does this work and the dump thread disappears immediately when STOP SLAVE is done on the slave?

I thought the dump thread would be waiting for new data to appear in the binlog, and not notice the slave connection is gone until it tries to send new data to it. Are you sure this wait will not sit waiting for some time until some timeout triggers somewhere?

Reply:
Well, that’s what happened for me, the dump thread starts terminating as soon as the replica disconnects (and, before this fix, leaves the `wait_condition.inc` with `Lost connection to server during query`).
It works in CI too.

Have you tried it out yourself?

Reply:
> Did you find a different behavior when you tried it out yourself?

Interesting, this is something specific to semi-sync. Indeed, I have seen slave dump threads hanging around previously, but that was non-semi sync. Now I learned that semi-sync on the slave explicitly connects to the master and does a KILL CONNECTION on its dump thread, which I did not know about:
```c++
void Repl_semi_sync_slave::kill_connection(MYSQL *mysql) {
...
  size_t kill_buffer_length = my_snprintf(kill_buffer, 30, "KILL %lu",
                                        mysql->thread_id);
  mysql_real_query(kill_mysql, kill_buffer, (ulong)kill_buffer_length);
```
This is apparently some AliBaba patch merged in 2017.

Reply:
👍 a(n undocumented?) part of [MDEV-13073](https://jira.mariadb.org/browse/MDEV-13073)